risedev.yml

```diff
@@ -128,6 +128,21 @@ profile:
       - use: compactor
       # - use: prometheus
       # - use: grafana
+  webhdfs:
+    steps:
+      # - use: etcd
+      - use: meta-node
+      - use: compute-node
+      - use: frontend
+      # If you want to use hdfs as storage backend, configure hdfs namenode and root path:
+      - use: opendal
+        engine: webhdfs
+        namenode: "127.0.0.1:9000"
```

Comment:
My mistake. webhdfs's config is differnt from hdfs, we should specify the endpoint to `http://127.0.0.1:9000`:

https://opendal.databend.rs/opendal/services/struct.Webhdfs.html

Reply:
And we need to config `delegation_token` as well?

Reply:
And we need to config `delegation_token` as well?

Reply:
> And we need to config `delegation_token` as well?

Not need for now. Most hdfs users don't setup this. We plan to support kerberos in the future.

Reply:
Seems that `hummock_001/2.data` is identified as  invalid path format🤔

Reply:
> Seems that `hummock_001/2.data` is identified as invalid path formatthinking

No, the path the valid. But `127.0.0.1:9000` seems not a valid http url.

Reply:
I have changed the endpoint to `"http://127.0.0.1:9870"`, but still the same error.

Reply:
I have changed the endpoint to `"http://127.0.0.1:9870"`, but still the same error.

Reply:
Maybe this is because the hadoop environment is not correct, let me take the time to check it.

Reply:
Maybe this is because the hadoop environment is not correct, let me take the time to check it.

Reply:
Can you output the debugging of the builder via `debug!("{builder:?}")` in `new_webhdfs_engine`? The error is returned by opendal and not related to hdfs itself.

Reply:
Seems that already connected to webhdfs.
```
2023-02-17T06:37:28.784007Z DEBUG opendal::services::webhdfs::backend: building backend: Builder { root: Some("rw"), endpoint: Some("127.0.0.1:9870") }    
2023-02-17T06:37:28.784111Z DEBUG opendal::services::webhdfs::backend: backend use root /rw/    
2023-02-17T06:37:28.784131Z DEBUG opendal::services::webhdfs::backend: backend use endpoint 127.0.0.1:9870    
2023-02-17T06:37:28.793976Z DEBUG opendal::services::webhdfs::backend: checking working directory: /rw/   
```

Reply:
Seems that already connected to webhdfs.
```
2023-02-17T06:37:28.784007Z DEBUG opendal::services::webhdfs::backend: building backend: Builder { root: Some("rw"), endpoint: Some("127.0.0.1:9870") }    
2023-02-17T06:37:28.784111Z DEBUG opendal::services::webhdfs::backend: backend use root /rw/    
2023-02-17T06:37:28.784131Z DEBUG opendal::services::webhdfs::backend: backend use endpoint 127.0.0.1:9870    
2023-02-17T06:37:28.793976Z DEBUG opendal::services::webhdfs::backend: checking working directory: /rw/   
```

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7718ef31167dd2c2a7062be0fb2c03d7

Reply:
I got it. It's a bug of OpenDAL that didn't prepend `http` for input endpoint. We will fix it in https://github.com/datafuselabs/opendal/issues/1353. You can workaround it via adding `http://` before endpoint.

Reply:
That is indeed a problem. But even if I hardcode the endpoint as `http://127.0.0.1:9870 `,  still the same error. 

```
 DEBUG opendal::services::webhdfs::backend: backend use endpoint http://127.0.0.1:9870  
```

```
 WARN risingwave_storage::hummock::compactor::shared_buffer_compact: Shared Buffer Compaction failed with error: ObjectStore failed with IO error Unexpected (permanent) at write, context: { service: webhdfs, service: webhdfs, path: hummock_001/1.data } => redirection fail, source: missing field `Location` at line 1 column 155.
```

Reply:
Oh, sorry.

webhdfs does not seem to handle rediection correctly, we will fix this in https://github.com/datafuselabs/opendal/issues/1355.

Reply:
The `namenode` has a port for HDFS and another port for HTTP, did this service use the same port?
And seems like the HTTP response lacks a `Location` field, what's the version of HDFS? For Hadoop3, an HTTP 200 OK response carrying a `Location` JSON should be given.

Reply:
hdfs 2.7.3

Reply:
> hdfs 2.7.3

Just checked the documentation and found out it does not support the `noredirect` parameter the services relying on. I'll fix this ASAP.